kt/entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/generator/hintstring/EnumFlagHintStringGenerator.kt

```diff
@@ -0,0 +1,12 @@
+package godot.entrygenerator.generator.hintstring
+
+import godot.entrygenerator.model.RegisteredProperty
+
+
+class EnumFlagHintStringGenerator(
+    registeredProperty: RegisteredProperty
+) : PropertyHintStringGenerator(registeredProperty) {
+
+    override fun getHintString(): String =
+        throw UnsupportedOperationException("Hint string for enums is handled by the binding at runtime.")
```

Comment:
Isn't that weird to have a hint that will always throw an error? It's kinda confusing for the users.
It's like "Ho yeah we support this feature but don't call it explicitly".

Reply:
This error should never get thrown in the first place. If the entry gen code works correctly, it should generate different code for enums which should never rely on hint string generators in the first place. These are generated at runtime here: https://github.com/utopia-rise/godot-kotlin-jvm/blob/ed18f34f64316d8e0c022eccca551531dff3f043/kt/godot-runtime/src/main/kotlin/godot/runtime/Registration.kt#L142
This exception is just here so that it fails fast if we change something in the entry gen and not handle enums correctly anymore.

Reply:
But it still seems weird to not use the GetHintString of a generator, just to generate it in another way inside the registration code.

Reply:
I had to earlier because it was not possible to just get the enum values the way our kotlin compiler plugin was set up before. We were also getting the properties and functions which might be defined inside the enum class.
So a reliable getting of just the enum values was not possible (at least in a easy to understand and maintainable way).
Not sure if this is still the case for ksp though. I'll look into it. Maybe we can merge the enumProperty and the normal property registration again.

Reply:
Ok it is possible but I'm not sure if it's worth it.
Because to be able to implement it i would need to add another callback param to the entry gen entry point here: https://github.com/utopia-rise/godot-kotlin-jvm/blob/7f979104bae5123a8a83ba99ade4073ac766a362/kt/entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/EntryGenerator.kt#L20
The purpose of the callback would be to resolve a `Type` to a `Clazz` so we can read the properties, functions and in the case of an enum; subclasses which are the enum values.
So the `godot-kotlin-symbol-processor` would need to keep track of the mapping of KSP types it mapped to entry gen `Type` so it can do that mapping when requested through the callback.
All of that is not complicated and shouldn't take too long to implement, but I'm not sure if it's worth it just for this case.
I'd rather just remove this hintstring provider entirely if it bothers you but if you think otherwise I'm fine with implementing this resolve function as well.

Reply:
Can't you store the information directly in the EnumFlagHintAnnotation, just like you do for some others like RangeHint.

Reply:
I can but i don't want to as then the user would have to write the names of the enums everytime by hand. Now it's done automatically. On top of that I would need to add additional checks to ensure the number of names match the number of enums in the case of enum, lists. Also i would need to check if no name is entered twice. Additionally i would need to implement this error reporting in the idea plugin as well.

Reply:
I don't get it. Can't you just get this information directly from the KSP without the user writing anything?

Currently, in the registration, you just take the Enum and iterate all the values to create the string, can't you simply move that part directly in the symbol processing?

Reply:
Yeah i could, but either I'd have to implement such a resolve callback like i mentioned earlier, or i cheat by enriching the `EnumHintAnnotation` model of the entry gen with values which are not actually present on the "real" annotation which i don't like to do beacause then the model does not represent the actual source code anymore.

So basically two options to do such a thing automatically during entry generation:
- Adding a resolve callback to the interface of the entry gen so calling tools like `godot-kotlin-symbol-processor` can implement it to resolve a `Type` to a `Clazz`.
- Or cheat by adding the needed information to one of the existing models to which it does not belong

Because the information about what enum values an enum has, is information on class level, so it should belong into the `Clazz` model.
But when checking the type of a property it is not of type `Clazz` it's of type `Type`. In compiler classes (or ksp for that matter) you then have a `resolve` method to get the class declaration. This is not present in my abstraction because it's not needed and has to be implemented by the calling tool (`godot-kotlin-symbol-processor` in the case of kotlin). So this would be the first option.

Reply:
The second option is the one I was thinking about. It's easy to implement and I don't see how it doesn't follow the Model. Imo, it's perfectly in line. We want what Godot considers to be an int with an EnumFlag hint. So there is nothing wrong with storing the different enum values inside the EnumFlagAnnotation.
Maybe the difference is in how we see the model. The actual implementation we use in Kotlin is a Set of Enum which is then converted back and forth to an Int but it only matters for Kotlin, another JVM language might not use the same kind of trick in the first place.  
The model should reflect the Godot side which is the reference, not the language source which can be interchanged by another one. We are the one cheating in the first place already by using a Set<Enum>.
I guess this is what you call cheating because the Kotlin source is a Set<Enum> property with an annotation without parameters. But the Godot side is an Int with an annotation holding multiples values (I say annotations in the context of godot both for the current export syntax in GDScript and the true annotations we will get in GDSCript 2.0 in Godot 4.0)

Reply:
True we already kinda cheat with the enum set. But the model does not cheat yet. The model is meant to be a abstracted representation of the actual source code. Which is the case atm.
If we add that additional information to the models `EnumHintAnnotation` or `EnumFlagHintAnnotation` it does not represent the actual source code anymore, because there that information is not present in the annotation.

That's what i mean by cheating. I just ask myself if it's worth it to let the model be different from the source code just for this one case.

So yeah i guess the difference is how we see the model. I saw it as a abstracted, language independent representation of the source code. And according to your comment, you rather see it as an conversion from the language to the godot side of things.

So i guess we first need to find a common ground what the model should represent. I'm still more keen to see it as an abstraction of the actual source code. And IMO it's the job of the entry gen to use that abstracted information to convert/generate it to something godot can understand and use.
So from my point of view, if we really want to change the enum registration, we should do it with a `resolve` function and not with a model change.

But of course I'm open to change my view on how we see the model. But atm i just don't see why it should represent the godot side and not the source code side of things

Reply:
To sum up and document the call @CedNaru, @piiertho and i just had:
We'll look at the model as an abstracted version of the things Godot needs, not like an AST.
One of the main reasons: stuff like `enumFlags` might be of different type in different languages and we cannot rely on that in the entry gen. Also it keeps the model more in line with changes to the GDScript exports (which will be annotations for Godot 4.0 as well).
So I'll change the enum hint string generation to store the enum values as metadata inside the anonotation in the model (option 2 mentioned above).
So we can get rid of the different registration of enum types.

The only thing that still will be registered differently, is the `enumFlag` type.
This one will be reworked once we have the entry gen type configuration mentioned below:

With a later PR, I'll make the entry gen configurable so it knows about built in types.
Currently fqNames of kotlin builtin types are hardcoded inside the entry gen. Also some other types are hardcoded.
These should be configurable when using the entry gen, so it knows what type is an Integer language independent

Reply:
done